File: topoflow/beliefs.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Discover Legal
"""Belief-delta rules (spec §10) [AF].

Heuristic, deterministic updates applied after a cell/generator runs. Small fixed
step sizes, clipped to [0,1]. handoff_quality is updated but excluded from the
signature (§4).
"""
from __future__ import annotations

from .config import Config
from .types import BeliefVector, CellOutput

STEP = 0.15  # default fixed step (0.1–0.2)
SMALL = 0.1


def _clip(x: float) -> float:
    return 0.0 if x < 0.0 else 1.0 if x > 1.0 else x


def role_of(skill: str) -> str:
    """Map a skill name to its §10 belief-source role category."""
    s = (skill or "").lower()
    if s.startswith("solver"):
        return "solver"
    if s == "planner":
        return "planner"
    if s == "memory" or s.startswith("web_search"):
        return "memory"
    if s.startswith("verifier"):
        return "verifier"
    if s == "synthesiser" or s == "synthesizer":
        return "synthesiser"
    if s == "critic":
        return "critic"
    if s == "evaluator":
        return "evaluator"
    return s


def _apply_role(b: BeliefVector, role: str, out: CellOutput, cfg: Config) -> BeliefVector:
    d = out.handoff_delta
    meta = out.meta
    if role == "planner" and d.get("subproblem") is not None:
        b.handoff_quality = _clip(b.handoff_quality + STEP)
        b.uncertainty = _clip(b.uncertainty - STEP)
    elif role == "memory":
        n = len(d.get("evidence") or []) if isinstance(d.get("evidence"), (list, tuple)) else (1 if d.get("evidence") else 0)
        b.evidence = _clip(b.evidence + min(0.3, SMALL * max(1, n)))
        b.uncertainty = _clip(b.uncertainty - SMALL)
        b.handoff_quality = _clip(b.handoff_quality + SMALL)
    elif role == "solver" and d.get("draft_answer") is not None:
        b.correctness = _clip(b.correctness + STEP)
        b.uncertainty = _clip(b.uncertainty - STEP)
        b.handoff_quality = _clip(b.handoff_quality + STEP)
    elif role == "critic" and d.get("critique") is not None:
        b.contradiction = _clip(b.contradiction + STEP)
        b.uncertainty = _clip(b.uncertainty + SMALL)
    elif role == "verifier":
        verdict = str(meta.get("verdict") or d.get("verification") or "").lower()
        if "support" in verdict or "pass" in verdict or "correct" in verdict:
            b.correctness = _clip(b.correctness + STEP)
            b.uncertainty = _clip(b.uncertainty - STEP)
            b.contradiction = _clip(b.contradiction - SMALL)
            b.evidence = _clip(b.evidence + SMALL)
        elif "refut" in verdict or "fail" in verdict or "incorrect" in verdict:
            b.correctness = _clip(b.correctness - STEP)
            b.contradiction = _clip(b.contradiction + STEP)
            b.uncertainty = _clip(b.uncertainty + SMALL)
    elif role == "synthesiser" and d.get("merged_answer") is not None:
        b.correctness = _clip(b.correctness + STEP)
        b.handoff_quality = _clip(b.handoff_quality + STEP)
        b.uncertainty = _clip(b.uncertainty - STEP)
    # evaluator: no belief delta (its decision feeds reward, not state)
    return b


def update_beliefs_cell(b: BeliefVector, out: CellOutput, cfg: Config) -> BeliefVector:
    return _apply_role(b, role_of(out.skill), out, cfg)


def update_beliefs_topology(b: BeliefVector, res, cfg: Config) -> BeliefVector:
    """For a composite cell [NEW]: derive the delta from which handoff fields its
    merged result populated and from its final verifier/tester outcome — i.e.
    treat the cell's net effect through the same §10 table, applied once."""
    fields = res.merged_handoff.fields
    if fields.get("merged_answer") is not None:
        b = _apply_role(b, "synthesiser", CellOutput(skill="synthesiser", handoff_delta=fields), cfg)
    elif fields.get("draft_answer") is not None:
        b = _apply_role(b, "solver", CellOutput(skill="solver", handoff_delta=fields), cfg)
    if fields.get("verification") is not None:
        b = _apply_role(
            b, "verifier",
            CellOutput(skill="verifier", handoff_delta=fields, meta={"verdict": fields.get("verification")}),
            cfg,
        )
    if fields.get("evidence") is not None:
        b = _apply_role(b, "memory", CellOutput(skill="memory", handoff_delta=fields), cfg)
    return b

File: biglaw-go/internal/topoflow/beliefs.go
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal

package topoflow

import "strings"

const beliefStep = 0.15 // default fixed step (0.1–0.2)
const beliefSmall = 0.1

// roleOf maps a skill name to its §10 belief-source role category.
func roleOf(skill string) string {
	s := strings.ToLower(skill)
	switch {
	case strings.HasPrefix(s, "solver"):
		return "solver"
	case s == "planner":
		return "planner"
	case s == "memory" || strings.HasPrefix(s, "web_search"):
		return "memory"
	case strings.HasPrefix(s, "verifier"):
		return "verifier"
	case s == "synthesiser" || s == "synthesizer":
		return "synthesiser"
	case s == "critic":
		return "critic"
	case s == "evaluator":
		return "evaluator"
	}
	return s
}

func applyRole(b BeliefVector, role string, delta map[string]any, meta map[string]any) BeliefVector {
	has := func(k string) bool { v, ok := delta[k]; return ok && v != nil }
	switch role {
	case "planner":
		if has("subproblem") {
			b.HandoffQuality = clamp01(b.HandoffQuality + beliefStep)
			b.Uncertainty = clamp01(b.Uncertainty - beliefStep)
		}
	case "memory":
		nb := 0
		if lst, ok := delta["evidence"].([]any); ok {
			nb = len(lst)
		} else if has("evidence") {
			nb = 1
		}
		inc := beliefSmall * float64(max1(nb))
		if inc > 0.3 {
			inc = 0.3
		}
		b.Evidence = clamp01(b.Evidence + inc)
		b.Uncertainty = clamp01(b.Uncertainty - beliefSmall)
		b.HandoffQuality = clamp01(b.HandoffQuality + beliefSmall)
	case "solver":
		if has("draft_answer") {
			b.Correctness = clamp01(b.Correctness + beliefStep)
			b.Uncertainty = clamp01(b.Uncertainty - beliefStep)
			b.HandoffQuality = clamp01(b.HandoffQuality + beliefStep)
		}
	case "critic":
		if has("critique") {
			b.Contradiction = clamp01(b.Contradiction + beliefStep)
			b.Uncertainty = clamp01(b.Uncertainty + beliefSmall)
		}
	case "verifier":
		verdict := strings.ToLower(asString(meta["verdict"]) + asString(delta["verification"]))
		switch {
		case strings.Contains(verdict, "support") || strings.Contains(verdict, "pass") || strings.Contains(verdict, "correct"):
			b.Correctness = clamp01(b.Correctness + beliefStep)
			b.Uncertainty = clamp01(b.Uncertainty - beliefStep)
			b.Contradiction = clamp01(b.Contradiction - beliefSmall)
			b.Evidence = clamp01(b.Evidence + beliefSmall)
		case strings.Contains(verdict, "refut") || strings.Contains(verdict, "fail") || strings.Contains(verdict, "incorrect"):
			b.Correctness = clamp01(b.Correctness - beliefStep)
			b.Contradiction = clamp01(b.Contradiction + beliefStep)
			b.Uncertainty = clamp01(b.Uncertainty + beliefSmall)
		}
	case "synthesiser":
		if has("merged_answer") {
			b.Correctness = clamp01(b.Correctness + beliefStep)
			b.HandoffQuality = clamp01(b.HandoffQuality + beliefStep)
			b.Uncertainty = clamp01(b.Uncertainty - beliefStep)
		}
	}
	// evaluator: no belief delta.
	return b
}

func updateBeliefsCell(b BeliefVector, out CellOutput) BeliefVector {
	return applyRole(b, roleOf(out.Skill), out.HandoffDelta, out.Meta)
}

// updateBeliefsTopology derives the delta from which handoff fields the merged
// result populated + the final verifier/tester outcome — the §10 table, once [NEW].
func updateBeliefsTopology(b BeliefVector, res TopologyResult) BeliefVector {
	f := res.MergedHandoff.Fields
	if f["merged_answer"] != nil {
		b = applyRole(b, "synthesiser", f, nil)
	} else if f["draft_answer"] != nil {
		b = applyRole(b, "solver", f, nil)
	}
	if f["verification"] != nil {
		b = applyRole(b, "verifier", f, map[string]any{"verdict": f["verification"]})
	}
	if f["evidence"] != nil {
		b = applyRole(b, "memory", f, nil)
	}
	return b
}

func max1(n int) int {
	if n < 1 {
		return 1
	}
	return n
}
